File: compiler/src/feedback/items/StructureFeedback.swift
func registerStructureFeedback(into context: FeedbackContext) {
    context.register(
        id: "missing-field",
        query: Queries.fact(MissingField.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, missing in
        writer.write(node)
        writer.write(" is missing a pattern for the field ")
        writer.write(code: missing.name)
        writer.write(".")
        writer.writeBreak()
        writer.write("Try adding a pattern for this field using ")
        writer.write(code: ":")
        writer.write(".")
    }

    context.register(
        id: "extra-field",
        query: Queries.fact(ExtraField.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, _, extra in
        writer.write(code: extra.name)
        writer.write(" isn't a field on this structure.")
        writer.writeBreak()
        writer.write("Double-check your spelling.")
    }

    context.register(
        id: "duplicate-field",
        query: Queries.fact(DuplicateField.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, _ in
        writer.write(node)
        writer.write(" is defined more than once.")
        writer.writeBreak()
        writer.write("Try removing this field.")
    }
}

File: compiler/feedback/src/items/structure_feedback.rs
use crate::{FeedbackCtx, FeedbackRank};
use wipple_queries::fact;
use wipple_syntax::expressions::structure_expression::{DuplicateField, ExtraField, MissingField};

pub fn register(ctx: &mut FeedbackCtx<'_>) {
    ctx.feedback("missing-field")
        .query(fact::<MissingField>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, MissingField(name)| {
            writer.node(node);
            writer.string(" is missing a pattern for the field ");
            writer.code(name);
            writer.string(".");
            writer.line_break();
            writer.string("Try adding a pattern for this field using ");
            writer.code(":");
            writer.string(".");
        })
        .register();

    ctx.feedback("extra-field")
        .query(fact::<ExtraField>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, _, extra| {
            writer.code(&extra.0);
            writer.string(" isn't a field on this structure.");
            writer.line_break();
            writer.string("Double-check your spelling.");
        })
        .register();

    ctx.feedback("duplicate-field")
        .query(fact::<DuplicateField>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" is defined more than once.");
            writer.line_break();
            writer.string("Try removing this field.");
        })
        .register();
}
